rfcs/proposed/flow_graph_fold_node/README.md

```diff
@@ -0,0 +1,226 @@
+If the node is processing the ``fold_stream_end``, it stores the end flag to ``true`` and check if all of the tasks were processed.
+The result would be considered complete and propagated to successors if both completion flag was set to ``true`` and all of the tasks were processed, i.e.
+the corresponding reference counter equals to _0_.
+
+## Process Specific Information
+
+### Propagating the stream index to the successors
+
+It may be useful for the successors of the ``fold_node`` to identify the input stream, the result for which was received:
+
+```cpp
+tbb::flow::graph g;
+
+tbb::flow::fold_node<int, int> fold(g, ...);
+tbb::flow::function_node<int, int> postprocess(g, tbb::flow::unlimited,
+    [](const tbb::flow::tagged_msg<std::size_t, int>& tagged_input) {
+        // tagged_input.tag() - a unique tag of the input stream
+        // cast_to<int>(tagged_input) - the fold result
+    });
+```
+
+One of the solutions is to provide flexibility:
+
+* If the ``OutputType`` template argument of the ``fold_node`` is specified as a specialization of ``tagged_msg`` - the tag should be propagated to the
+successors. In this case the successors should be ready to take the same ``tagged_msg`` (be a ``receiver<tagged_msg<...>>``).
+* If the ``OutputType`` template argument of the ``fold_node`` is not a specialization of ``tagged_msg`` - the tag should *not* be propagated and the
+successors are not expected to be receivers of ``tagged_msg``.
+
+### Propagation of the stream index to the node body
+
+Same as in the previous question, there may be useful for some use-cases to propagate the index of the input stream to the fold operation to
+implement the specific logic for different stream tags:
+
+```cpp
+tbb::flow::graph g;
+
+using fold_node_type = fold_node<int, int>;
+using input_type = typename fold_node_type::input_type;
+
+fold_node_type fold(g, tbb::flow::unlimited,
+    /*fold initializer = */0,
+    [](int lhs, const input_type& rhs) {
+        if (rhs.tag() == 0) {
+            // Fold logic for tag 0
+            return lhs + cast_to<int>(rhs);
+        } else {
+            // Fold logic for any other streams
+            return (lhs / 2) + (cast_to<int>(rhs) / 2);
+        }
+    });
+```
+
+The same flexible approach can be applied for solving this as well:
+* If the invocation of _fold operation_ with an object of type
+
+It is also an open question, should ``input_type`` be allowed as a first, second, or both operands of _fold operation_.
+
+Similar to the previous open question, the flexible approach can be applied to allow both use-cases:
+* If the invocation of the _fold operation_ associated with the ``fold_node`` with the argument of type ``input_type`` (i.e. ``tagged_msg``)
+is well-formed - propagate the index of the stream to the body.
+* Otherwise - the index of the input stream is not propagated to the _fold operation_.
+
+### Should ``InputType`` and ``OutputType`` be different?
```

Comment:
I think yes, generally these should be different types. If I remember correctly, one of our motivating scenarios is computing a histogram for a data sequence; it is obviously of a different type than the data elements.